homeassistant/components/tplink_omada/__init__.py

```diff
@@ -51,22 +54,31 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         ) from ex
 
     site_client = await client.get_site_client(OmadaSite("", entry.data[CONF_SITE]))
-    controller = OmadaSiteController(hass, site_client)
+    entry.runtime_data = controller = OmadaSiteController(hass, site_client)
     gateway_coordinator = await controller.get_gateway_coordinator()
     if gateway_coordinator:
         await gateway_coordinator.async_config_entry_first_refresh()
     await controller.get_clients_coordinator().async_config_entry_first_refresh()
 
-    hass.data[DOMAIN][entry.entry_id] = controller
-
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     return True
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_unload_entry(hass: HomeAssistant, entry: OmadaConfigEntry) -> bool:
     """Unload a config entry."""
-    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
-        hass.data[DOMAIN].pop(entry.entry_id)
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def async_remove_config_entry_device(
+    hass: HomeAssistant, config_entry: OmadaConfigEntry, device_entry: DeviceEntry
+) -> bool:
+    """Remove a config entry from a device."""
+    controller = config_entry.runtime_data
+
+    devices = await controller.omada_client.get_devices()
 
-    return unload_ok
+    # Allow removal if the device is not present in the controller
+    return not any(
+        mac for _, mac in device_entry.identifiers if mac in (d.mac for d in devices)
+    )
```

Comment:
Sure, I'll take a look. I copied what Unifi was doing, as it's going to be pretty rare that networking devices are removed. But I probably should be registering the devices explicitly instead of doing it implicitly in the entities.

Reply:
Creating devices explicitly is nothing wrong btw, if it works for you, it works.